collector/src/compare.rs

```diff
@@ -0,0 +1,110 @@
+use database::{Connection, Lookup};
+use itertools::Itertools;
+use tabled::{Table, Tabled};
+
+const DEFAULT_SIGNIFICANCE_THRESHOLD: f64 = 0.002;
+
+/// Compare 2 artifacts and print the result.
+pub async fn compare_artifacts(mut conn: Box<dyn Connection>, base: String, modified: String) {
+    let index = database::Index::load(&mut *conn).await;
+    let sids = index
+        .compile_statistic_descriptions()
+        .map(|(_, sid)| sid)
+        .collect::<Vec<_>>();
+
+    let base_id_number = conn
+        .artifact_by_name(&base)
+        .await
+        .expect("Cannot find specified artifact")
+        .lookup(&index)
+        .unwrap();
+    let modified_id_number = conn
+        .artifact_by_name(&modified)
+        .await
+        .expect("Cannot find specified artifact")
+        .lookup(&index)
+        .unwrap();
+
+    let pstats = conn
+        .get_pstats(&sids, &[Some(base_id_number), Some(modified_id_number)])
+        .await;
+    let tuple_pstats = pstats
+        .into_iter()
+        .map(|row| row.into_iter().collect_tuple::<(_, _)>().unwrap())
+        .collect::<Vec<(Option<f64>, Option<f64>)>>();
+
+    #[derive(Tabled)]
+    struct Regression {
+        count: usize,
+        #[tabled(display_with = "display_range")]
+        range: (f64, f64),
+        #[tabled(display_with = "display_mean")]
+        mean: f64,
+    }
+
+    fn display_range(value: &(f64, f64)) -> String {
+        format!("[{:+.2}%, {:+.2}%]", value.0, value.1)
+    }
+
+    fn display_mean(value: &f64) -> String {
+        format!("{:+.2}%", value)
+    }
+
+    impl From<&Vec<f64>> for Regression {
+        fn from(value: &Vec<f64>) -> Self {
+            let min = *value.iter().min_by(|a, b| a.total_cmp(b)).unwrap();
+            let max = *value.iter().max_by(|a, b| a.total_cmp(b)).unwrap();
+            let count = value.len();
+
+            Regression {
+                range: (min, max),
+                count,
+                mean: (value.iter().sum::<f64>() / count as f64),
+            }
+        }
+    }
+
+    let change = tuple_pstats
+        .iter()
+        .filter_map(|&(a, b)| match (a, b) {
+            (Some(a), Some(b)) => {
+                if a == 0.0 {
```

Comment:
Comparing floats with zero is.. tricky :laughing: But I suppose in this case we can just let it be.